### PR TITLE
STT-1697 : Feature/GoogleAuth -> develop

### PR DESCRIPTION
**Pull request by Lapatelli**
[*] Startup: Added Authentication Middleware
[*] All: Property "Telephone Number" has been removed
[*] ChatApp.API.Controllers:
-Added AuthorizeAttribute
-Added AccountController: Login and LogOut using GoogleApi+
   ChatController:
-Routes have been renamed;
-Added mapping to Chat.ChatUsers so as to hide unnecessary information
   UserController:
-Added new controller (GetAllChatsForUserAsync)
[*] ChatApp.API.Mapper:
-Added necessary mapping;
[*] ChatApp.API.ViewModels:
-Added ChatInfoViewModel, UserInfoViewModel, CreateChatViewModel, RegisterViewModel, UserViewModel
-Added property Picture to Chat View Models;
-Added property BytePhoto to User View Models;
[*] ChatApp.Core:
-DTO and Entities have been transformed so as to work with MongoDB correctly;
[*] ChatApp.CQRS:
-Added RegisterUserCommand, SetStatusCommand and GetUserByEmailQuery to CQRS
-Added LeaveChatHandler and LeaveChatCommand;
-Added GetAllChatsForUserHandler and GetAllChatsForUserQuery;
-Added new Queries to Db for get User by Email and Updating UserCollection
-Refactored queries, commands, handlers;
-Added folder "Shared" for testing approach to share common functionality;
[*] ChatApp.Interfaces:
-Added new methods to IChat and IUser Repositories;
[*] ChatApp.Persistence:
-Added methods SaveChandes and AddCommand to ChatDbContext for query's consistency and realization of pattern "UnitOfWork"
-ChatRepository and UserRepository methods have been changed for realiztion of "Left Join" and code cleaning;

### CODE REVIEW

File: ChatApp/ChatApp.API/Controllers/AccountController.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            var user = await _mediator.Send(new RegisterUserCommand(model.FirstName, model.LastName, model.EmailAddress, model.Photo));
+
+            var result = _mapper.Map<User, UserViewModel>(user);
+
+            return Ok(result);
+        }
+
+        [HttpGet]
+        [Route("google-login")]
```
Comment (dmyhun-andersen) [inline ChatApp/ChatApp.API/Controllers/AccountController.cs]: [HttpGet("google-login")]

File: ChatApp/ChatApp.API/Controllers/AccountController.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                GoogleDefaults.AuthenticationScheme,
+                new AuthenticationProperties
+                {
+                    RedirectUri = Url.Action("signin-google", new { ReturnUrl = returnUrl })
+                });
+        }
+
+        [HttpGet("signin-google")]
+        public async Task<IActionResult> LoginCallback(string returnUrl = null)
+        {
```
Comment (dmyhun-andersen) [inline ChatApp/ChatApp.API/Controllers/AccountController.cs:56]: try this
public async Task<IActionResult> LoginCallback(string returnUrl = "~/")

File: ChatApp/ChatApp.API/Controllers/AccountController.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            returnUrl = returnUrl ?? Url.Content("~/");
+            
+            var authenticateResult = await HttpContext.AuthenticateAsync();
+
+            if (!authenticateResult.Succeeded)
+                return BadRequest();
+
+            string email = authenticateResult.Principal.FindFirstValue(ClaimTypes.Email);
```
Comment (dmyhun-andersen) [inline ChatApp/ChatApp.API/Controllers/AccountController.cs]: use **var** where it is possible.

Comment (dmyhun-andersen) [reply ChatApp/ChatApp.API/Controllers/AccountController.cs]: probably in this case **string** is ok

File: ChatApp/ChatApp.API/Mapper/ChatProfile.cs [hunk 4]
```diff
@@ -16,77 +17,94 @@ public ChatProfile()
...
+            CreateMap<ChatDTO, ChatInfoViewModel>()
+                .ConvertUsing(src => new ChatInfoViewModel
                 {
-                    Id = ObjectId.Parse(src.chatId),
-                    Name = src.model.Name,
-                    Password = src.model.Password,
-                    ChatPrivacy = src.model.ChatPrivacy
+                    Id = src.Id.ToString(),
+                    Name = src.Name,
+                    //Password = src.Password,
```
Comment (dmyhun-andersen) [inline ChatApp/ChatApp.API/Mapper/ChatProfile.cs]: It's better to remove commented code

File: ChatApp/ChatApp.API/Startup.cs [hunk 5]
```diff
@@ -28,16 +32,34 @@ public Startup(IConfiguration configuration)
...
+               .AddCookie(options =>
+                   {
+                       options.LoginPath = new PathString("/Account/google-login");
+                       options.Cookie.HttpOnly = true;
+                   }
+               )
+               .AddGoogle(options =>
+               {
+                   options.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+                   options.ClientId = "158339939419-r3gh8lgsd3dp7nl1kihedga6q44080gk.apps.googleusercontent.com";
```
Comment (dmyhun-andersen) [inline ChatApp/ChatApp.API/Startup.cs]: It's better to move ClientId and ClientSecret to json config file.

File: ChatApp/ChatApp.Core/Entities/User.cs [hunk 6]
```diff
@@ -1,19 +1,11 @@
-﻿using ChatApp.Core.Enums;
+﻿using ChatApp.Core.DTO;
 using System.Collections.Generic;
 
 namespace ChatApp.Core.Entities
 {
-    public class User
+    public class User : UserDTO
```
Comment (dmyhun-andersen) [inline ChatApp/ChatApp.Core/Entities/User.cs]: I am not sure that it is good decision when domain entity inherits from data transfer object.

File: ChatApp/ChatApp.API/Controllers/AccountController.cs (hunk 1 above)
Comment (Lapatelli) [reply ChatApp/ChatApp.API/Controllers/AccountController.cs]: Done!

File: ChatApp/ChatApp.API/Controllers/AccountController.cs (hunk 2 above)
Comment (Lapatelli) [reply ChatApp/ChatApp.API/Controllers/AccountController.cs:56]: Done!

File: ChatApp/ChatApp.API/Controllers/AccountController.cs (hunk 3 above)
Comment (Lapatelli) [reply ChatApp/ChatApp.API/Controllers/AccountController.cs]: Done!

File: ChatApp/ChatApp.API/Mapper/ChatProfile.cs (hunk 4 above)
Comment (Lapatelli) [reply ChatApp/ChatApp.API/Mapper/ChatProfile.cs]: Done!

File: ChatApp/ChatApp.API/Startup.cs (hunk 5 above)
Comment (Lapatelli) [reply ChatApp/ChatApp.API/Startup.cs]: Done!

File: ChatApp/ChatApp.Core/Entities/User.cs (hunk 6 above)
Comment (Lapatelli) [reply ChatApp/ChatApp.Core/Entities/User.cs]: Done!

**Pull request merged**